src/broker/boot_flub.c

```diff
@@ -0,0 +1,223 @@
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* boot_flub.c - FLUx Boot protocol
+ *
+ * Add a broker to an existing Flux instance.
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <errno.h>
+#include <jansson.h>
+#include <flux/core.h>
+
+#include "src/common/libidset/idset.h"
+#include "src/common/libutil/errprintf.h"
+#include "src/common/libutil/errno_safe.h"
+#include "src/common/libutil/ipaddr.h"
+#include "ccan/str/str.h"
+
+#include "attr.h"
+#include "overlay.h"
+#include "topology.h"
+
+#include "boot_flub.h"
+
+struct boot_info {
+    int size;
+    int rank;
+    json_t *attrs;
+};
+
+struct boot_parent {
+    char *pubkey;
+    int rank;
+    const char *uri;
+};
+
+static int wait_for_group_membership (flux_t *h,
+                                      const char *name,
+                                      int rank,
+                                      flux_error_t *error)
+{
+    flux_future_t *f;
+    bool online = false;
+
+    if (!(f = flux_rpc_pack (h,
+                             "groups.get",
+                             0,
+                             FLUX_RPC_STREAMING,
+                             "{s:s}",
+                             "name", "broker.online"))) {
+        errprintf (error, "broker.online: %s", strerror (errno));
+        return -1;
+    }
+    do {
+        const char *members;
+        struct idset *ids;
+
+        if (flux_rpc_get_unpack (f, "{s:s}", "members", &members) < 0) {
+            errprintf (error, "broker.online: %s", future_strerror (f, errno));
+            break;
+        }
+        if ((ids = idset_decode (members))
+            && idset_test (ids, rank))
+            online = true;
+        idset_destroy (ids);
+        flux_future_reset (f);
+    } while (!online);
+    flux_future_destroy (f);
+
+    return online ? 0 : -1;
+}
+
+static int set_attrs (attr_t *attrs, json_t *dict)
+{
+    const char *key;
+    json_t *val;
+
+    json_object_foreach (dict, key, val) {
+        const char *s = json_string_value (val);
+        if (!s) {
+            errno = EPROTO;
+            return -1;
+        }
+        if (attr_add (attrs, key, s, ATTR_IMMUTABLE) < 0)
+            return -1;
+    }
+    return 0;
+}
+
+int boot_flub (struct broker *ctx, flux_error_t *error)
+{
+    const char *uri = NULL;
+    flux_t *h;
+    flux_future_t *f = NULL;
+    flux_future_t *f2 = NULL;
+    struct topology *topo = NULL;
+    const char *topo_uri;
+    struct boot_info info;
+    struct boot_parent parent;
+    const char *bind_uri = NULL;
+    int rc = -1;
+
+    /* Ask a Flux instance to allocate an available rank.
+     * N.B. the broker unsets FLUX_URI so either it is set on the
+     * command line via broker.boot-server, or the compiled-in path
+     * is used (system instance).
+     */
+    (void)attr_get (ctx->attrs, "broker.boot-server", &uri, NULL);
+    if (!(h = flux_open_ex (uri, 0, error)))
```

Comment:
Any reason not to call `uri_resolve(3)` here? Then a JOBID or other resolvable URI could be provided directly to the `broker.boot-server` attribute.

Although, if the intent is to add a higher level command like `flux expand OPTIONS... JOBID`, perhaps that's unnecessary.

Reply:
It seems like this might be helpful.  I'm not sure what the next steps are for tooling so may as well add that for now.

Reply:
A complication is that `uri_resolve(3)` requires FLUX_URI to be set, and the broker unsets that.  I could setenv FLUX_URI to the value of the `parent-uri` attribute around the call but since the environment is global...eww?  Maybe we should table this one for now.

Reply:
Hm, yeah, that's a bummer. It might be useful in the long term in case you target the URI of a job that hasn't started yet, but doesn't seem like a big deal for now.

Maybe we need a version of `uri_resolve(3)` that takes a `uri` argument and (somehow) passes that down to `flux uri` if that command is invoked. Meh, probably too soon to worry about it.